### ISSUE TITLE
Bluewave Support

### ISSUE THREAD

**Issue by ghost** [NuSkooler/enigma-bbs#119]
#31 

I've been doing some thinking about this, seems Tiny mentioned it would be good to have. Bluewave and QWK share some things in common, and I'm thinking that maybe the message system might need some additions to support it. The two main things, I am thinking will be needed are an area tag for each message area, and some kind of subscription system.

The area tag could just be another variable in the config.hjson for each message area, but the subscription thing, I'm not sure about. It could be useful for things like new message scans as well as QWK and blue wave to determine which areas a user is interested in. Of course the subscription thing could be confined to a bluewave or qwk (or bluewave/qwk) module, (or completley left out and just have a user get everything), but I think a more generic approach would be better, as it would make any other future modules able to use it.

With those two things, I think everything else could just be a module that scans areas the user is interested in, packs the messages into a file and sends it to the user. (and the other way for uploading). 

I'm not sure how to go about a subscription thing, but would be willing to have a crack at making a module to do bluewave things if you want.

**Comment by shighfield** [comment]
Even if it's the way you did Magicka... download everything.  It's all a good place to start.   ;)

**Comment by apamment** [comment]
> ENiGMA treats messages as it's own, and any other format is "external". If you look at the way FTN is implemented, it follows this model (provide import and/or export functionality). QWK/others can be the same way.

So, you're saying the QWK/Bluewave exporting/importing should be a scanner tosser module, it could then be invoked by some sort of menu module? I like that, you could then use the same QWK exporting/importing module for QWK networking, just with a different frontend.

> Messages already belong to an area and contain a local area tag. Again, in the case of external (e.g. FTN) you'll see that a mapping occurs (external area <> internal); this is only a byproduct of FTN but not a requirement, but of course the same system can be used elsewhere.

Yep, I think (QWK at least) has an 8 character limitation on area names, so I think you'd want to use that mapping,

**Comment by NuSkooler** [comment]
> So, you're saying the QWK/Bluewave exporting/importing should be a scanner tosser module, it could then be invoked by some sort of menu module? I like that, you could then use the same QWK exporting/importing module for QWK networking, just with a different frontend.

Yup exactly. The `MessageScanTossModule` class may need a little love to make it on-demand callable vs on schedules/etc. only, but that's super easy to do. 

For an example of user prefs, perhaps look at the file area config. You can basically store a JSON object of any type in a user pref -- subscribed areas in this case. ...and finally, a `config.hjson` section maps external (QWK/Bluewave/...) to local areas.

**Comment by NuSkooler** [comment]
For reference (if anyone knows of other/updated refs please share!):
http://help-site.com/local/BWSTRUCT.TXT

**Comment by shighfield** [comment]
As far as I can remember that is the last version of the structures 2.01, I looked in my old bp directory and that's what I have as well from back then when I was attempting to write a web based BBS using CGI programs. (It kind of worked ;) )

**Comment by jhenkins** [comment]
Any movements on this? It would be great to be able to do the offline reading thing, since MultiMail has solid support for BlueWave.

To add to @NuSkooler's   list of BW reference material above, I stumbled across some bits on William McBrine's site that to my untrained eye looks as if it might be newer than 2.0.1:

http://wmcbrine.com/mmail/specs/bwdev300.tar.gz

The boilerplate in bluewave.h has the following version line:

Version 3 - November 30, 1995 

I hope this is useful in some way.

**Comment by NuSkooler** [comment]
@jhenkins I haven't done any work on this yet - this is lower on the totem than a number of other things currently (biggest are currently: WFC > re-write FSE code).

If someone wants to take a stab at this stuff, ENiG's messages are already pluggable which is how FTN already works. NNTP, Gopher and some others show read-only, etc. so there is a good amount of example code as to how to achieve it.

**Comment by NuSkooler** [comment]
Please see the notes in #31 RE: QWK. 

After testing/etc., if there is enough want for Bluewave specific, I may look into it.

**Comment by andy5995** [comment]
I started working on this in #798 and #799

**Comment by NuSkooler** [comment]
Export landed in #798 (55bb48a7) — a caller can take a Blue Wave packet away and read it offline, reached with <kbd>B</kbd> from the message base menu.

Reading reply packets back in is tracked separately in #810, so closing this one on the export half. Eight and a half years is long enough for a ticket to sit open. Thanks @andy5995.
